### ISSUE TITLE
map over object keys

### ISSUE THREAD

**Issue by indeyets** [ramda/ramda#1361]
I'm looking for  way to map over object keys. Something like:

```javascript
let obj = {a: 1, b: 2} 
let result = mapKeys(k => `${k}z`, obj)  // {az: 1, bz: 2}
```

didn't find anything like that by browsing thru the docs

**Comment by kwijibo** [comment]
maybe something like 
`compose(fromPairs,map(mapFunc), toPairs)`
?

**Comment by CrossEye** [comment]
I don't know any immediate way to do that outside actually writing that function (presumably [modeled on][sc] [`R.mapObj`][mo]:

```js
var mapKeys = R.curry(function mapKeys(fn, obj) {
  return R.reduce(function(acc, key) {
    acc[fn(key)] = obj[key];
    return acc;
  }, {}, R.keys(obj));
});

mapKeys(k => `${k}z`, {a: 1, b: 2});  //=> {az: 1, bz: 2}
```

I've never used anything like this.  How will you use it?

  [mo]: http://ramdajs.com/docs/#mapObj 
  [sc]: https://github.com/ramda/ramda/blob/master/src/mapObj.js#L29-L34

**Comment by asaf-romano** [comment]
As I noted elsewhere, I've my own variant of this function in our code base, which renames a particular property of an object, leaving everything else as-is. It's quite useful for "translating" objects from 3rd-party models into the local model.

**Comment by thecotne** [comment]
here is my implementation for `mapKeys`

```js
import {
    curry,
    keys,
    map,
    values,
    zipObj
} from 'ramda';

export let mapKeys = curry((fn, obj) => zipObj(map(fn, keys(obj)), values(obj)));
```

**Comment by CrossEye** [comment]
See #1124, #1129, #1137 for other implementations.

This one was a close call for inclusion in the library; eventually it was decided that this belonged in the [Cookbook][co], but it looks as though no one ever did actually include it.

  [co]: https://github.com/ramda/ramda/wiki/Cookbook

**Comment by raine** [comment]
> See #1124, #1129, #1137 for other implementations.

Those are not actually the same function.

The idea with `objFromKeys` is `(a → b) → [a] → {a: b}`.

It's in the [Cookbook](https://github.com/ramda/ramda/wiki/Cookbook) now.

**Comment by CrossEye** [comment]
Hey, you're right, they're not!

They say that memory's the first thing to go.  I forget what the second one is...

Thanks.

**Comment by smmoosavi** [comment]
Why this issue is closed? All of this solution casing a hard time for typescript.

**Comment by CrossEye** [comment]
@smmoosavi:

What issues are you having?  Do note that while we like to make things smoother for Typescript users, we don't generally go far out of our way for it, as this is a JS library and at last check it used features not compatible with TS typings.  That may since have changed (I've heard rumors) but it's not been a major motivating factor.

**Comment by zatziky** [comment]
Hey guys, can you please add the `mapKeys` to the lib?  🙏 
We need to map the keys while keeping the same values.

It's a second time this year we need it and on a different project.
Lodash provides it out-of-the-box by the way.

**Comment by CrossEye** [comment]
@zatziky:

----------
**Note**
I've done this before, and swore I would make sure never to do it again.  I wrote a medium-length response to this yesterday, thought I'd posted it, but must have left it in preview mode without posting.  Then somehow I killed that tab, and it's gone.

I probably won't go into as much detail this time around; I'm a little frustrated.

----------

If you want to create a PR for this, I promise it will get a fair hearing.  But there are no promises that it will be accepted.  Ramda has to balance between not over-expanding its API and providing useful functions.  So a PR would need to explain either why this is hard to do correctly in user code or why its so common that Ramda should simply supply it.

BTW, what lodash -- or Underscore, or any other libraries -- offer is not much of a source for Ramda.  We are more inspired by other FP languages than other JS libraries, but we don't treat either the languages or the libraries as guidelines, only inspiration.  Conversely, we would not expect libraries like lodash to support every feature we offer.  I don't think they have `lift` or transducers or lenses, and I would not expect them to.

----------

I have been thinking about creating a more general function that would make `mapKeys` a very simple function to write:

```js
const rebuild = (fn) => (obj) =>
  fromPairs (chain (apply (fn)) (toPairs (obj)))

const remap = (fn) =>
  rebuild ((k, v) => [fn (k, v)])

const mapKeys = (fn) =>
  remap ((k, v) => [fn (k), v])

mapKeys (k => `${k}z`) ({a: 1, b: 2}) //=> {az: 1, bz: 2}

rebuild (
  (k, v) => v % 3 == 0
    ? [[`${k}1`, v / 3], [`${k}2`, v / 3], [`${k}3`, v / 3]]
  : v % 3 == 1
    ? [[k, v]]
  : []
) ({a: 8, b: 6, c: 7, d: 5, e: 3, f: 0, g: 9})
//=> { b1: 2, b2: 2, b3: 2, c: 7, e1: 1, e2: 1, e3: 1, 
//     f1: 0, f2: 0, f3: 0, g1: 3, g2: 3, g3: 3 }
```

`rebuild` would take a function from a key and value to an array of key-value pairs, and an object, split that object into key-value pairs, call the function on each and reconstitute the result into an object.  You can see in the final call the power of `rebuild`, where we take the key-values and, depending on the value, return three, one, or no new key-values for our final object.

`remap` is a simple gloss on this for those cases when there's a one-to-one mapping between input key-value pairs and output ones.

With that `mapKays` is simple to write.  Whether that would then be included in Ramda, I don't know.  I feel though that `rebuild` at least is a likely candidate, and possibly `remap`.

Of course this is also easy to write in modern JS:

```js
const rebuild = (fn) => (obj) =>
  Object .fromEntries (Object .entries (obj) .flatMap (e => fn (...e)))
```

so we'd have to have the same discussion about the benefits and drawbacks of that function.

But if we included `rebuild` and `remap`, would you still think a `mapKeys` function would be important?

**Comment by zatziky** [comment]
Hey Scott, @CrossEye 

> If you want to create a PR for this, I promise it will get a fair hearing. But there are no promises that it will be accepted.

It's better to first know something will be accepted before creating a PR. 
I already a created PR while ago that was not accepted because you want to keep the ops limited.
So, I won't put my time to risk again. ;-) 

> I have been thinking about creating a more general function that would make `mapKeys` a very simple function to write

That's exactly 1 of 3 main issues I have with Ramda. 
I am aware we can build such methods as `mapKeys` ourselves. 
But we should be given common operators out-of-the-box.
Ramda has awesome documentation and we could nicely leverage it.
See, you just type in **map** and get all relevant functions.
You even see if they are **dedicated to a list or an object**.

![image](https://user-images.githubusercontent.com/2794365/108475172-a68c4600-7290-11eb-9725-e3fff05c60c6.png)

That said bulk of operators should not be an issue if you can quickly find what you need.

**Comment by CrossEye** [comment]
@zatziky:

>> If you want to create a PR for this, I promise it will get a fair hearing. But there are no promises that it will be accepted.
> It's better to first know something will be accepted before creating a PR.

I have no problem continuing to discuss without a PR, but it's often easier to have something specific to discuss.

> I already a created PR while ago that was not accepted because you want to keep the ops limited.
> So, I won't put my time to risk again. ;-)

Sorry if you had a bad experience.  I've never looked to see what portion of PRs we accept, but I find them the best way to play around with enhancements, to actually try them and share them.  That does mean they have to be considered experimental until approved.

>> I have been thinking about creating a more general function that would make mapKeys a very simple function to write

> That's exactly 1 of 3 main issues I have with Ramda.

What are the other two?

> I am aware we can build such methods as mapKeys ourselves.
> But we should be given common operators out-of-the-box.

Ramda has to balance several concerns.  We can't bloat the library with every function one might ever want.  This is not only a matter of file size, but also one of conceptual weight.  And even excellent documentation searching is not the solution.  We're not likely to find `mapKeys` when looking for `renameProperties`.  And while I'd love to build a [Hoogle][ho]-like signature search, that probably wouldn't help for many problems.

But we want to offer the common functions, and especially the common ones that are hard to write oneself.  The idea is to be a useful, helpful library, with a design aimed at a specific problem: making it easier to write functional code in JS.  

We've never figured out a definitive answer on what makes a function a good candidate for Ramda.  We don't include `isArray`, `isObject`, `isFunction`, or `isString`, finding that a common `is` is clean enough, and still easily allow the user to write those they need:

```js
const isString = is (String)
const isArray = is (Array)
// later
if (isString (response)) {
  // handle string
} else if (isArray (response)) {
  // handle array
} else {
  // handle error
}
```

(Of course, I usually don't bother, simply doing `is (String) (response)` or `is (String, response)`.)

But we do include both `compose` and `pipe`, even though either would be easy enough to derive from the other.  It seemed to the team that each was so useful that it was worth having both.

So we have to make a judgment call on these, based on a few factors: how common the requirement seems to be, how close the function seems to Ramda's core mission, and how easy or difficult it would be for users to write correctly on their own.

So new functions not proposed by the core team or without strong immediate support from a team member tend to need a champion, someone who can argue those factors on behalf of the candidate function.

I have almost never needed a `mapKeys` function, so I'm not likely to be the champion of it.  But there are other functions that have been included in Ramda that I or my teammates rarely use; and someone successfully argued for them, either in issues or as pull requests.  That you've needed this function twice in the recent year is a useful data point, but not compelling enough on its own.

In other words, we're willing to be convinced, but you need to actually do the convincing.

  [ho]: https://hoogle.haskell.org/

**Comment by ericfreed-pki** [comment]
The requested functionality is what `mapObjIndexed` already provides, no?

From the documentation:
```
const prependKeyAndDouble = (num, key, obj) => key + (num * 2);
R.mapObjIndexed(prependKeyAndDouble, { x: 1, y: 2, z: 3 }); //=> { x: 'x2', y: 'y4', z: 'z6' }
```
[REPL](https://ramdajs.com/repl/?v=0.22.1#?const%20vIn%20%3D%20%7B%20x%3A%201%2C%20y%3A%202%2C%20z%3A%203%20%7D%3B%0Aconst%20prependKeyAndDouble%20%3D%20%28num%2C%20key%2C%20obj%29%20%3D%3E%20key%20%2B%20%28num%20%2A%202%29%3B%0A%0AR.mapObjIndexed%28prependKeyAndDouble%2C%20vIn%29%3B%20%2F%2F%3D%3E%20%7B%20x%3A%20%27x2%27%2C%20y%3A%20%27y4%27%2C%20z%3A%20%27z6%27%20%7D)

**Comment by CrossEye** [comment]
@ericfreed-pki:  That's a different function.  `mapObjIndexed` allows us to consider the values of the related keys when mapping the values, but it doesn't allow us to change keys.  The idea of `mapKeys` is different, allowing us to change the *keys*:

```js
mapKeys (toUpper, {foo: 1, bar: 2}) //=> {FOO: 1, BAR: 2}
```
